ambry-store/src/main/java/com.github.ambry.store/StoreMessageReadSet.java

```diff
@@ -13,96 +13,127 @@
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * A read option class that maintains the offset and size
  */
-class BlobReadOptions implements Comparable<BlobReadOptions> {
-  private final Long offset;
+class BlobReadOptions implements Comparable<BlobReadOptions>, Closeable {
+  private final LogSegment segment;
+  private final Pair<File, FileChannel> segmentView;
+  private final Offset offset;
   private final Long size;
   private final Long ttl;
   private final StoreKey storeKey;
-  private Logger logger = LoggerFactory.getLogger(getClass());
+  private final AtomicBoolean open = new AtomicBoolean(true);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
 
   private static final short version = 0;
   private static final short Version_Length = 2;
   private static final short Offset_Length = 8;
   private static final short Size_Length = 8;
   private static final short TTL_Length = 8;
 
-  BlobReadOptions(long offset, long size, long ttl, StoreKey storeKey) {
+  BlobReadOptions(Log log, Offset offset, long size, long ttl, StoreKey storeKey) {
+    segment = log.getSegment(offset.getName());
+    if (offset.getOffset() + size > segment.getEndOffset()) {
+      throw new IllegalArgumentException(
+          "Invalid offset [" + offset + "] and size [" + size + "]. Segment end offset: " + "[" + segment.getEndOffset()
+              + "]");
+    }
+    segmentView = segment.getView();
     this.offset = offset;
     this.size = size;
     this.ttl = ttl;
     this.storeKey = storeKey;
     logger.trace("BlobReadOption offset {} size {} ttl {} storeKey {}", offset, size, ttl, storeKey);
   }
 
-  public long getOffset() {
-    return offset;
+  long getOffset() {
+    return offset.getOffset();
   }
 
-  public long getSize() {
+  long getSize() {
     return size;
   }
 
-  public long getTTL() {
+  long getTTL() {
     return ttl;
   }
 
-  public StoreKey getStoreKey() {
+  StoreKey getStoreKey() {
     return storeKey;
   }
 
-  public MessageInfo getMessageInfo() {
+  MessageInfo getMessageInfo() {
     return new MessageInfo(storeKey, size, ttl);
   }
 
-  public boolean validateFileEndOffset(long fileEndOffset) {
-    return offset + size <= fileEndOffset;
+  File getFile() {
+    return segmentView.getFirst();
+  }
+
+  FileChannel getChannel() {
+    return segmentView.getSecond();
   }
 
   @Override
   public int compareTo(BlobReadOptions o) {
-    return offset.compareTo(o.getOffset());
+    return offset.compareTo(o.offset);
   }
 
-  public byte[] toBytes() {
+  byte[] toBytes() {
     byte[] buf = new byte[Version_Length + Offset_Length + Size_Length + TTL_Length + storeKey.sizeInBytes()];
     ByteBuffer bufWrap = ByteBuffer.wrap(buf);
     bufWrap.putShort(version);
-    bufWrap.putLong(offset);
+    bufWrap.putLong(offset.getOffset());
```

Comment:
for version 0, we have a long at this position. but for version 1, we will have an object of "Offset" class right. So this might need fixing or am i missing something